src/base/rss/rss_autodownloader.cpp

```diff
@@ -0,0 +1,374 @@
+}
+
+QList<AutoDownloadRule> AutoDownloader::rules() const
+{
+    return m_rules.values();
+}
+
+void AutoDownloader::setRule(const AutoDownloadRule &rule)
+{
+    if (!m_rules.contains(rule.name())) {
+        // Insert new rule
+        setRule_impl(rule);
+        emit ruleAdded(rule.name());
+        m_dirty = true;
+    }
+    else if (m_rules[rule.name()] != rule) {
+        // Update existing rule
+        setRule_impl(rule);
+        emit ruleChanged(rule.name());
+        m_dirty = true;
+    }
+
+    store();
+}
+
+bool AutoDownloader::renameRule(const QString &ruleName, const QString &newRuleName)
+{
+    if (!m_rules.contains(ruleName)) return false;
+    if (m_rules.contains(newRuleName)) return false;
+
+    m_rules.insert(newRuleName, m_rules.take(ruleName));
+    emit ruleRenamed(newRuleName, ruleName);
+    m_dirty = true;
+    store();
+    return true;
+}
+
+void AutoDownloader::removeRule(const QString &ruleName)
+{
+    if (m_rules.remove(ruleName) > 0) {
+        emit ruleRemoved(ruleName);
+        m_dirty = true;
+    }
+
+    store();
+}
+
+void AutoDownloader::process()
+{
+    if (m_processingQueue.isEmpty()) return; // was be disabled
+
+    processJob(m_processingQueue.takeFirst());
+    if (!m_processingQueue.isEmpty())
+        // Schedule to process the next torrent (if any)
+        m_processingTimer->start();
+
+    store();
+}
+
+void AutoDownloader::handleTorrentDownloadFinished(const QString &url)
+{
+    auto job = m_waitingJobs.take(url);
+
+    auto feed = Session::instance()->feedByURL(job->feedURL);
+    if (feed) {
+        if (auto article = feed->articleByGUID(job->articleGUID))
+            article->markAsRead();
+    }
+}
+
+void AutoDownloader::handleTorrentDownloadFailed(const QString &url)
+{
+    m_waitingJobs.remove(url);
+    // TODO: Re-schedule job here.
+}
+
+void AutoDownloader::handleNewArticle(Article *article)
+{
+    if (!article->isRead() && !article->torrentUrl().isEmpty())
+        addJobForArticle(article);
+}
+
+void AutoDownloader::setRule_impl(const AutoDownloadRule &rule)
+{
+    m_rules.insert(rule.name(), rule);
+}
+
+void AutoDownloader::addJobForArticle(Article *article)
+{
+    QSharedPointer<ProcessingJob> job(new ProcessingJob);
+    job->feedURL = article->feed()->url();
+    job->articleGUID = article->guid();
+    job->articleTitle = article->title();
+    job->torrentURL = article->torrentUrl();
+    m_processingQueue.append(job);
+    if (!m_processingTimer->isActive())
+        m_processingTimer->start();
+}
+
+void AutoDownloader::processJob(const QSharedPointer<ProcessingJob> &job)
+{
+    foreach (const QString &ruleName, m_rules.keys()) {
+        AutoDownloadRule &rule = m_rules[ruleName];
+        if (!rule.isEnabled()) continue;
+        if (!rule.feedURLs().contains(job->feedURL)) continue;
+        if (!rule.matches(job->articleTitle)) continue;
+
+//        rule.setLastMatch(QDateTime::currentDateTime());
```

Comment:
I believe the comment would be placed better right where the question arises. Although, in my code I usually use `#ifdef` in such cases:
```c++
#if USE_OPTIMIZED_CODE_FOR_FEATURE_X
 blah-blah optimized
#else
 vanilla blah-blag
#endif
```